### ISSUE TITLE
Spot instance fails to initialize

### ISSUE THREAD

**Issue by jinfanzon** [npalm/terraform-aws-gitlab-runner#680]
# Describe the bug
I've deployed the module couple of months ago and was working with no problem until this morning all my gitlab pipelines were failing. Checking the runner logs I can see:

`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: {     "driver": "amazonec2",     "level": "error",     "msg": "Error creating machine: Error running provisioning: ssh command error:",     "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",     "operation": "create",     "time": "2023-02-02T14:21:17Z" }`
`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: {     "driver": "amazonec2",     "level": "error",     "msg": "command : printf '%s' '-----BEGIN CERTIFICATE-----",     "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",     "operation": "create",     "time": "2023-02-02T14:21:17Z" }`
`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: 
{
    "error": "exit status 1",
    "fields.time": 74004702268,
    "level": "error",
    "msg": "Machine creation failed",
    "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",
    "time": "2023-02-02T14:21:17Z"
}`
`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: {     "driver": "amazonec2",     "level": "error",     "msg": "' \| sudo tee /etc/docker/ca.pem",     "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",     "operation": "create",     "time": "2023-02-02T14:21:17Z" }`
`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: {     "driver": "amazonec2",     "level": "error",     "msg": "err     : exit status 1",     "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",     "operation": "create",     "time": "2023-02-02T14:21:17Z" }`
`Feb  2 14:21:17 ip-10-0-1-50 gitlab-runner: {     "driver": "amazonec2",     "level": "error",     "msg": "output  : tee: /etc/docker/ca.pem: No such file or directory",     "name": "runner-zned3iq9-kalto-runners-1675347603-20f902b1",     "operation": "create",     "time": "2023-02-02T14:21:17Z" }`

# To Reproduce

1. Apply terraform Module
2. Run gitlab pipeline using the runner
3. You'll notice a spot instance is in state initializing and then shut downs

# Expected behavior

Spot instance should initialize with no issues and run the build

# Additional context

tf module settings:

````
module "runner" {
  source  = "npalm/gitlab-runner/aws"
  version = "5.9.0"
  aws_region = var.aws_region
  environment = var.environment
  vpc_id                   = module.vpc.vpc_id
  runners_use_private_address = false
  subnet_ids_gitlab_runner = module.vpc.public_subnets
  subnet_id_runners = element(module.vpc.public_subnets, 0)
  metrics_autoscaling = ["GroupDesiredCapacity", "GroupInServiceCapacity"]

  runner_instance_spot_price= "on-demand-price"
  runners_request_concurrency= 5
  runners_name = var.runner_name
  runners_monitoring = true
  runners_gitlab_url = var.gitlab_url
  enable_runner_ssm_access = true
  gitlab_runner_security_group_ids = [data.aws_security_group.default.id]
  runners_request_spot_instance = true
  docker_machine_spot_price_bid = "on-demand-price"
  docker_machine_instance_type = "t3.medium"
  instance_type            = "t3.small"
  gitlab_runner_registration_config = {
    registration_token = var.registration_token
    tag_list = "docker_spot_runner,aws,gitlab-org-docker,gitlab-org"
    description = "AWS runner - Spot"
    locked_to_project = "true"
    run_untagged = "false"
    maximum_timeout = "3600"
  }

  tags = {
    "tf-aws-gitlab-runner:shared" = "runner-default"
    "tf-aws-gitlab-runner:instancelifecycle" = "spot:yes"
  }

  runners_privileged = "true"
  runners_additional_volumes = ["/certs/client"]

  runners_volumes_tmpfs = [
    {
      volume = "/var/opt/cache",
      options = "rw,noexec"
    }
  ]

  runners_services_volumes_tmpfs = [
    {
      volume = "/var/lib/mysql",
      options = "rw,noexec"
    }
  ]
  runners_machine_autoscaling = [
    {
      periods    = ["\"* * 0-8 * * mon-sun *\""]
      idle_count = 0
      idle_time  = 60
      timezone   = "America/Argentina/Buenos_Aires"
    }
  ]

  runner_root_block_device = {
    volume_type = "standard"
    volume_size = 8
  }
}
````

**Comment by rpearson-catdevnull** [comment]
We've also just started experiencing this issue specifically in `eu-central-1`. We haven't seen this issue occur (yet!) in other regions.

**Comment by jhagestedt** [comment]
same for us also in eu-central-1.

**Comment by gordianberger** [comment]
Seeing this in eu-central-2 as well.

**Comment by rpearson-catdevnull** [comment]
Correction - we've now observed this same error for newly created instance in `eu-west-1`. (We just hadn't had any new instances spawned for a while).

**Comment by stefan-caraiman** [comment]
Related issue: https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/97 

TLDR: Seems to be due to recent Docker releases.

**Comment by rpearson-catdevnull** [comment]
We are currently trialling the following change to our terraform for this module:

`
  docker_machine_download_url = "https://gitlab-docker-machine-downloads.s3.amazonaws.com/mr-102/docker-machine-Linux-x86_64"
`

Based on this: https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/98#note_1263469511

Will report back shortly on progress

**Comment by xposix** [comment]
I'm having the same issue in eu-west-2 too, thanks for looking into it.

**Comment by gordianberger** [comment]
This works for us now:

`      "docker_machine_options": [
        "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'"
      ],`

**Comment by jinfanzon** [comment]
> We are currently trialling the following change to our terraform for this module:
> 
> `docker_machine_download_url = "https://gitlab-docker-machine-downloads.s3.amazonaws.com/mr-102/docker-machine-Linux-x86_64"`
> 
> Based on this: https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/98#note_1263469511
> 
> Will report back shortly on progress

This workaround does not seems to work:

Feb  2 15:56:16 ip-10-0-1-114 gitlab-runner: {
    "level": "error",
    "msg": "InvalidInstanceID.Malformed: The instance ID '' is malformed",
    "name": "runner-v-yvp9vg-kalto-runners-1675353373-976d2774",
    "operation": "stop",
    "time": "2023-02-02T15:56:16Z"
}

**Comment by rpearson-catdevnull** [comment]
@jinfanzon yeah its not worked for us either. We've been able to get instances up with that MR102 but then jobs just fail. We're trying https://github.com/npalm/terraform-aws-gitlab-runner/issues/680#issuecomment-1413964540 now

**Comment by Razorr1996** [comment]
It works for us

https://github.com/npalm/terraform-aws-gitlab-runner/issues/680#issuecomment-1413964540

**Comment by BehbudSh** [comment]
Any update on this issue? It stopped working hour ago. I'm using an On-Demand instance

**Comment by mbuotidem** [comment]
this also appears to work:

```
locals {
  user_data = <<EOT
#cloud-boothook
#!/bin/bash -ex
mkdir /etc/docker
EOT

}
```
```
module "gitlab-runner" {
  runners_use_private_address = false
  source                      = "npalm/gitlab-runner/aws"
  version                     = **"5.9.0"**    
  runners_userdata = local.user_data
  ...
  }

```

**Comment by BehbudSh** [comment]
> This works for us now:
> 
> ` "docker_machine_options": [ "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'" ],`

This works for us as well. Thanks!

**Comment by rpearson-catdevnull** [comment]
https://github.com/npalm/terraform-aws-gitlab-runner/issues/680#issuecomment-1413964540 also works for us (after some strange behaviour with the first instance)

**Comment by ermiaqasemi** [comment]
We are also experiencing the same issue in the `us-east-1` region.

**Comment by NK74** [comment]
@jinfanzon I experiencing the same, but i got a "MissingParameter InvalidInstanceID"

```
# docker-machine ls
NAME                           ACTIVE      DRIVER      STATE   URL   SWARM   DOCKER    ERRORS
runner-xxxxxxxxxxxx   -        amazonec2   Error       Unknown   MissingParameter: The request must contain the parameter InstanceId
```

**Comment by ermiaqasemi** [comment]
@NK74 I got the same issue, however, @Razorr1996 solution fixes the problem.

**Comment by jinfanzon** [comment]
This is the only [workaround](https://github.com/npalm/terraform-aws-gitlab-runner/issues/680#issuecomment-1414023055) that worked for me

**Comment by ermiaqasemi** [comment]
It seems they fixed the issue. https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/98#note_1263483883

**Comment by NK74** [comment]
Ouch, it doesn't work sadly :/ 


user-data on the main runner which refers to the docker-machine install
```
curl -L https://packages.gitlab.com/install/repositories/runner/gitlab-runner/script.rpm.sh | bash
yum install gitlab-runner-${gitlab_runner_version} -y

curl -L https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/releases/v0.16.2-gitlab.19/downloads/docker-machine-Linux-x86_64 >/tmp/docker-machine && \
  chmod +x /tmp/docker-machine && \
  cp /tmp/docker-machine /usr/local/bin/docker-machine && \
  ln -s /usr/local/bin/docker-machine /usr/bin/docker-machine
  
 # docker-machine version
docker-machine version 0.16.2-gitlab.19, build 12b41948
```

config.toml
```
[[runners]]
  name = "xxxxxxx"
  ....
  executor = "docker+machine"
  [runners.custom_build_dir]
  [runners.cache]
    Type = "s3"
    Shared = true
    MaxUploadedArchiveSize = 0
    [runners.cache.s3]
      ServerAddress = "s3.amazonaws.com"
      BucketName = "xxxxxx"
      BucketLocation = "eu-west-1"
  [runners.docker]
    tls_verify = false
    image = "docker:20.10.21"
    privileged = true
    disable_entrypoint_overwrite = false
    oom_kill_disable = false
    disable_cache = false
    volumes = [
        "/var/run/docker.sock:/var/run/docker.sock", 
        "/root/.docker:/root/.docker", 
        "/cache"
    ]
    shm_size = 0
  [runners.machine]
    IdleCount = 1
    IdleScaleFactor = 0.0
    IdleCountMin = 0
    IdleTime = 600
    MachineDriver = "amazonec2"
    MachineName = "runner-%s"
    MachineOptions = [
        "amazonec2-ami=ami-00e7df8df28dfa791", 
        "engine-install-url=https://releases.rancher.com/install-docker/20.10.21.sh", 
        "amazonec2-instance-type=c5.xlarge", 
        "amazonec2-region=eu-west-1", 
        "amazonec2-vpc-id=vpc-xxxxxx", 
        "amazonec2-zone=a", 
        "amazonec2-subnet-id=subnet-xxxxxxx", 
        "amazonec2-private-address-only=true", 
        "amazonec2-request-spot-instance=true", 
        "amazonec2-spot-price=0.20", 
        "amazonec2-security-group=main-runner-docker-machine", 
        "amazonec2-tags=xxxxxxx", 
        "amazonec2-monitoring=0", 
        "amazonec2-iam-instance-profile=main-runner-runners-profile", 
        "amazonec2-root-size=60", 
        "amazonec2-userdata=/etc/gitlab-runner/user-data.sh"
    ]
    [[runners.machine.autoscaling]]
        Periods = ["* * * * * sat,sun *"]
        IdleCount = 0
        IdleTime = 600
        Timezone = "Europe/Paris"
```



gitlab-runner register
```
Runtime platform                                    arch=amd64 os=linux pid=2973 revision=12335144 version=15.8.0
Running in system-mode.

WARNING: Support for registration tokens and runner parameters in the 'register' command has been deprecated in GitLab Runner 15.6 and will be replaced with support for authentication tokens. For more information, see https://gitlab.com/gitlab-org/gitlab/-/issues/380872
Registering runner... succeeded                     runner=jcq_7RAf
Runner registered successfully. Feel free to start it, but if it's running already the config should be automatically reloaded!

```

docker-machine ls
```
NAME           ACTIVE   DRIVER      STATE   URL   SWARM   DOCKER    ERRORS
runner-xxxxxx   -        amazonec2   Error                 Unknown   MissingParameter: The request must contain the parameter InstanceId
```

tail /var/log/messages
```
WARNING: This action will delete both local reference and remote instance.#033[0;m  name#033[0;m=runner-xxxxxxx operation#033[0;m=remove(runner-xxxxxxx) Missing instance ID, this is likely due to a failure during machine creation#033[0;m  name#033[0;m=runner-xxxxxxx operation#033[0;m=remove
Successfully removed runner-xxxxxxx#033[0;m  name#033[0;m=runner-xxxxxxx operation#033[0;m=remove#033[0;33mWARNING: Problem while reading command output     #033[0;m  #033[0;33merror#033[0;m=read |0: file already closed#033[0;33mWARNING: Problem while reading command output     #033[0;m  #033[0;33merror#033[0;m=read |0: file already closed
Machine removed                                   #033[0;m  lifetime#033[0;m=2.98151711s name#033[0;m=runner-xxxxxxx now#033[0;m=2023-02-02 18:23:31.37437395 +0000 UTC m=+573.171060881 reason#033[0;m=Failed to create retries#033[0;m=0 used#033[0;m=233.007865ms usedCount#033[0;m=0
```

**Comment by kayman-mk** [comment]
Thanks to @bwmetcalf version 5.9.1 of the module was released a minute ago. Check it out. It uses the newest docker+machine version which should fix the issue.

**Comment by NK74** [comment]
I'm already on the latest 0.16.2-gitlab.19

**Comment by mattjtodd** [comment]
Using `eu-west-2` and seeing the same issues with docker-machine.

Error message which from machine which seems relevant is

`Error creating machine: Error in driver during machine creation: Error request spot instance: InvalidParameter: 'instance' is not a valid taggable resource type for this operation.`

**Comment by nathan-gilbert** [comment]
I'm seeing this as well in `us-west-1`

**Comment by NK74** [comment]
Runners report this issue in sentry (it may help 🤷‍♂️ )
_MissingParameter: The request must contain the parameter InstanceId_

```
*errors.errorString: Machine creation failed
  File "/builds/gitlab-org/gitlab-runner/helpers/sentry/log_hook.go", line 54, in (*LogHook).Fire
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/hooks.go", line 28, in LevelHooks.Fire
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go", line 272, in (*Entry).fireHooks
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go", line 241, in (*Entry).log
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go", line 293, in (*Entry).Log
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go", line 383, in (*Entry).Logln
  File "/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go", line 412, in (*Entry).Errorln
  File "/builds/gitlab-org/gitlab-runner/executors/docker/machine/provider.go", line 125, in (*machineProvider).createWithGrowthCapacity
  File "/builds/gitlab-org/gitlab-runner/executors/docker/machine/provider.go", line 105, in (*machineProvider).create.func1
  File "/builds/gitlab-org/gitlab-runner/executors/docker/machine/coordinator.go", line 44, in (*runnerMachinesCoordinator).waitForGrowthCapacity
```

**Comment by kayman-mk** [comment]
There is still a bug in the docker machine. See #681 and https://gitlab.com/gitlab-org/gitlab-runner/-/issues/29213#note_1263789121

5.9.1 fixes the issue for on demand instances. Spot instances still fail.

**Comment by nathan-gilbert** [comment]
I am running with spot instances, I'm on `5.9.1`, using the `docker_machine_options` above. Seeing this in my agent logs:

```json
{
  "driver": "amazonec2",
  "level": "info",
  "msg": "(runner-43hkterr-linux-small-1675371263-ac45de06) Missing instance ID, this is likely due to a failure during machine creation",
  "name": "runner-43hkterr-linux-small-1675371263-ac45de06",
  "operation": "create",
  "time": "2023-02-02T20:54:24Z"
}
```

```json
{
  "driver": "amazonec2",
  "level": "error",
  "msg": "Error creating machine: Error in driver during machine creation: Error request spot instance: InvalidParameter: 'instance' is not a valid taggable resource type for this operation.",
  "name": "runner-43hkterr-linux-small-1675371263-ac45de06",
  "operation": "create",
  "time": "2023-02-02T20:54:26Z"
}
```

```json
{
  "level": "error",
  "msg": "InvalidInstanceID.Malformed: The instance ID '' is malformed",
  "name": "runner-43hkterr-linux-small-1675371263-ac45de06",
  "operation": "stop",
  "time": "2023-02-02T20:54:26Z"
}
```

**Comment by nathan-gilbert** [comment]
I'm able to get to a working instance with spot instances by adding the following:

```terraform
docker_machine_version        = "0.16.2-gitlab.15"
docker_machine_options = [
    "engine-install-url='https://releases.rancher.com/install-docker/20.10.sh'"
]
```

**Comment by Nightbr** [comment]
@nathan-gilbert your workaround fix our spot instance on eu-west-3, thanks a lot :pray:

**Comment by NK74** [comment]
@nathan-gilbert Could you show us the .toml file ? 🙏

**Comment by nathan-gilbert** [comment]
I'm not sure it's working completely, I don't think they are getting cleaned up correctly. 

But here is my redacted toml:
```toml
oncurrent = 10
check_interval = 3
sentry_dsn = ""
log_format = "json"
listen_address = ""

[[runners]]
  name = "linux"
  url = "****"
  clone_url = ""
  token = "***"
  executor = "docker+machine"
  environment = []
  pre_build_script = ""
  post_build_script = ""
  pre_clone_script = ""
  request_concurrency = 1
  output_limit = 4096
  limit = 5
  [runners.docker]
    tls_verify = false
    image = "docker:18.03.1-ce"
    privileged = true
    disable_cache = false
    volumes = ["/cache","/certs/client"]
    extra_hosts = []
    shm_size = 0
    pull_policy = ["always"]
    runtime = ""
    helper_image = ""

  [runners.docker.tmpfs]
    "/var/opt/cache" = "rw,noexec"
  [runners.docker.services_tmpfs]
    "/var/lib/mysql" = "rw,noexec"
  [runners.cache]
    Type = "s3"
    Shared = true
    [runners.cache.s3]
      AuthenticationType = "iam"
      ServerAddress = "s3.amazonaws.com"
      BucketName = ""
      BucketLocation = "us-west-1"
      Insecure = false
  [runners.machine]
    IdleCount = 0
    IdleTime = 600
    MaxBuilds = 100
    MachineDriver = "amazonec2"
    MachineName = "linux-small-%s"
    achineOptions = [
      "amazonec2-instance-type=t3.small",
      "amazonec2-region=us-west-1",
      "amazonec2-zone=a",
      "amazonec2-vpc-id=",
      "amazonec2-subnet-id=",
      "amazonec2-private-address-only=true",
      "amazonec2-use-private-address=false",
      "amazonec2-request-spot-instance=true",
      "amazonec2-spot-price=",
      "amazonec2-security-group=",
      "amazonec2-tags=Environment,linux-small,ITAR,false,tf-aws-gitlab-runner:instancelifecycle,spot:yes,gitlab-runner-parent-id,i-0c402796cdae1d547",
      "amazonec2-use-ebs-optimized-instance=true",
      "amazonec2-monitoring=true",
      "amazonec2-iam-instance-profile=linux-small-docker-machine",
      "amazonec2-root-size=16",
      "amazonec2-volume-type=gp2",
      "amazonec2-userdata=",
      "amazonec2-ami=ami-03e0029f86a2c74c3"
      ,"amazonec2-metadata-token=required", "amazonec2-metadata-token-response-hop-limit=2","engine-install-url='https://releases.rancher.com/install-docker/20.10.sh'"
    ]

   [[runners.machine.autoscaling]]
    Periods = ["* * 0-9,17-23 * * mon-fri *","* * * * * sat,sun *"]
    IdleCount = 0
    IdleTime = 60
    Timezone = "America/Denver"
   [[runners.machine.autoscaling]]
    Periods = ["* * 9-17 * * mon-fri *"]
    IdleCount = 1
    IdleTime = 1800
    Timezone = "America/Denver"
```

**Comment by PhakornKiong** [comment]
@nathan-gilbert your workaround fix our spot instance on ap-southeast-1, thanks a lot 🙏

**Comment by lmzbonack** [comment]
@nathan-gilbert fix has also fixed our on demand runners in us-east-1. Thank you!

**Comment by rickychew77** [comment]
```
"docker_machine_options": [ "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'" ]

```
@gordianberger @Razorr1996 above works too but wondering if you using the latest module version `5.9.1` ? 
seems like `0.16.2-gitlab.19` causing another error from latest module version.

**Comment by CarlosPalaciosC** [comment]
This configuration fixes our spot instances in us-east-1:
```
MachineOptions = 
[
...
 ,"amazonec2-metadata-token=required", "amazonec2-metadata-token-response-hop-limit=2","engine-install-url='https://releases.rancher.com/install-docker/20.10.sh'"
]
```
Thank you @nathan-gilbert

**Comment by CarlosPalaciosC** [comment]
> ```
> "docker_machine_options": [ "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'" ]
> ```
> 
> @gordianberger @Razorr1996 above works too but wondering if you using the latest module version `5.9.1` ? seems like `0.16.2-gitlab.19` causing another error from latest module version.

If you are deploying in AWS, you could test:

```
MachineOptions = 
[
...
 ,"amazonec2-metadata-token=required", "amazonec2-metadata-token-response-hop-limit=2","engine-install-url='https://releases.rancher.com/install-docker/20.10.sh'"
]
```

**Comment by Razorr1996** [comment]
> ```
> "docker_machine_options": [ "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'" ]
> ```
> 
> 
>      
>   
> 
> @gordianberger @Razorr1996 above works too but wondering if you using the latest module version `5.9.1` ?
> seems like `0.16.2-gitlab.19` causing another error from latest module version.

5.9.1 doesn't work at all for us. With fix docker_machine_options or without.
I rolled back to 5.9.0.

**Comment by kayman-mk** [comment]
@Razorr1996 You need an old version of the Runner too. See https://github.com/npalm/terraform-aws-gitlab-runner/issues/680#issuecomment-1414372838

**Comment by javidrazack** [comment]
@nathan-gilbert Thanks man! you workaround fixed the spot instance issue in ap-south-1.

**Comment by nathan-gilbert** [comment]
I've updated my runner agent to use Gitlab Runner version `15.8.1` now and jobs appear to working correctly, though I'm seeing some error like this in the logs:
```json
{
  "builds": 0,
  "error": "exit code 1",
  "executor": "docker+machine",
  "level": "warning",
  "msg": "Failed to process runner",
  "runner": "43HKTerZ",
  "time": "2023-02-04T16:40:10Z"
}
```

Maybe those are expected? 

Spot instances seem to be cleaning themselves up as expected.

**Comment by Physium** [comment]
Understand the workaround to use an older version/enginer but is there any news on a fix for this?

**Comment by NK74** [comment]
I have the impression that there are several errors related to this subject.

On my side, the gitlab runner version does not affect (15.7.3 or 15.8.1 for example)
The update of the "engine-install-url" has indeed fixed the problem without touching anything else...

During my test i saw a bunch of errors in my sentry + a lot of runner instance (runner agent) unable to finish the deployment... Am i alone 😶‍🌫️ ?
<img width="1118" alt="image" src="https://user-images.githubusercontent.com/17825769/217083508-4ed9339f-8cea-435a-a087-f858374bb93b.png">

**Comment by DnD-Magnum** [comment]
This workaround doesn't work on my side. 
I have again the error message `gitlab-runner: {"level":"error","msg":"InvalidInstanceID.Malformed: The instance ID '' is malformed"`

Nevertheless, I use your workaround: 
```terraform
  docker_machine_options = [
    "amazonec2-metadata-token=required",
    "amazonec2-metadata-token-response-hop-limit=2",
    "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'",
  ]
  docker_machine_version = "0.16.2-gitlab.15"
  runners_image          = "public.ecr.aws/docker/library/docker:20.10.23-dind"

```

We are using: 
* EC2 Spot instance (on-demande-price)
* Ami: ami-0a89a7563fc68be84 
* Zone: eu-west-3
* npalm/gitlab-runner/aws module: 5.9.1
* Gitlab-Runner: 15.8.1

Ideas to make it work?

**Comment by kayman-mk** [comment]
I am using the module in version 5.9.1 configured as below What I discovered: Make sure that `runners_idle_count` is greater than zero. Nothing was working with a configuration value of zero!

```hcl
locals {
  gitlab_runner_version = "15.1.1"
}

module "gitlab_runner_test" {
  for_each = data.aws_subnet.runners
  source  = "npalm/gitlab-runner/aws"
  version = "5.9.1"

  aws_region                       = var.aws_region
  environment                      = format("%s-TEST-%s", local.resource_prefix, substr(each.value.availability_zone, -1, 1))
  vpc_id                           = var.vpc_id
  subnet_ids_gitlab_runner         = [each.value.id] # the agent is placed in the same subnet of the runners
  subnet_id_runners                = each.value.id
  docker_machine_instance_type     = "t3.micro"
  docker_machine_options           = ["engine-registry-mirror=https://my.mirror.redacted", "engine-install-url=https://releases.rancher.com/install-docker/20.10.21.sh"]
  docker_machine_version           = "0.16.2-gitlab.15"
  runners_name                     = format("%s-Gitlab-Runner-TEST-%s", local.resource_prefix, upper(substr(each.value.availability_zone, -1, 1)))
  enable_runner_ssm_access         = true
  enable_docker_machine_ssm_access = true
  enable_asg_recreation            = true
  runners_gitlab_url               = var.gitlab_url
  runners_image                    = "public.ecr.aws/docker/library/docker:20.10.23-dind" #"docker-proxy.sales-platform.hlag.com/automation/build-java8:stable"
  runners_idle_count               = 1
  runners_idle_time                = 1800
  runners_limit                    = 1  
  runners_concurrent               = 20 
  runners_max_builds               = 50
  runners_root_size                = 16
  runners_machine_autoscaling = [
    {
      periods    = ["\"* * 0-23 * * mon-sun *\""]
      idle_count = 3
      idle_time  = 2700
      timezone   = "Europe/Berlin"
  }]
  runners_output_limit        = 40960
  runners_request_concurrency = 1
  runners_additional_volumes  = ["/var/run/docker.sock:/var/run/docker.sock"]
  runners_pre_clone_script    = local.pre_clone_scripts_command
  userdata_post_install = templatefile("${path.module}/templates/runners_user_install_script",
    {
      gitlab_runner_version = local.gitlab_runner_version
  })
  runners_helper_image  = "redacted#gitlab-runner-helper-runner-14-8-3:stable"
  gitlab_runner_version = local.gitlab_runner_version
  runners_clone_url     = "ssh://git@gitlab.my.instance" # redacted

  cloudwatch_logging_retention_in_days = 5

  tags        = { for k, v in local.tags : k => v if k != "environment" }
  agent_tags  = local.tags
  runner_tags = local.tags
  role_tags   = tomap({ "allowReadPlatformCredentials" : "gitlab-ci-runner" })
  kms_key_id  = var.encryption_key_arn

  cache_shared = true

  cache_bucket = {
    create = false
    policy = "${module.cache.policy_arn}"
    bucket = "${module.cache.bucket}"
  }

  overrides = {
    name_sg                     = ""
    name_iam_objects            = ""
    name_runner_agent_instance  = "Glab-Ag-TEST-${each.value.availability_zone}"
    name_docker_machine_runners = "Glab-Exec-TEST-${each.value.availability_zone}"
  }

  runners_volumes_tmpfs = [
    {
      volume  = "/var/opt/cache",
      options = "rw,noexec"
    }
  ]

  gitlab_runner_registration_config = {
    registration_token = aws_ssm_parameter.gitlab_runner_registration_token.value
    tag_list           = "test"
    description        = "TEST (${each.value.availability_zone})"
    locked_to_project  = "false"
    run_untagged       = "false"
    maximum_timeout    = "10800"
  }
}
```

**Comment by pandarouxbsd** [comment]
Hi, 

I tried on the eu-central-1 region with the given configuration and it's not working neither with spot instance nor with on demand.

In the logs, I still have the following error

`"InvalidInstanceID.Malformed: The instance ID '' is malformed"`

Configuration:

```
docker_machine_version = "0.16.2-gitlab.15"
  docker_machine_options = [
    "engine-install-url='https://releases.rancher.com/install-docker/20.10.sh'"
  ]

  runners_request_spot_instance = false

  runners_idle_count = 1
  runners_idle_time  = 1800
  runners_limit      = 1
  runners_concurrent = 20
  runners_max_builds = 50
```

**Comment by gmeligio** [comment]
I don't use spot instances, but it's working for me with this configuration.

```terraform
version = "~> 5.9.1"
gitlab_runner_version = "15.8.2"
```

I removed `docker_machine_version` and `docker_machine_options`, and I'm leaving the defaults of the module.

**Comment by halaoner** [comment]
Deploying the solution in to `eu-central-2` region with the following configuration does not work either (neither work with `docker_machine_options = [
      "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'"
    ]`):

- `runner module version = 5.9.1`
- `instance_type = m6gd.medium`
- `docker_machine_instance_type = m6gd.medium`
- `runner_ami_filter = ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20230208`
- `ami_filter = amzn2-ami-kernel-5.10-hvm-2.0.20230119.1-arm64-gp2`
- `docker_machine_download_url = https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/releases/v0.16.2-gitlab.19/downloads/docker-machine-Linux-aarch64`

```bash
$ docker-machine --version
docker-machine version 0.16.2-gitlab.19, build 12b41948
```

Terraform configuration:

```terraform
module "runner" {
  # https://registry.terraform.io/modules/npalm/gitlab-runner/aws/
  source  = "npalm/gitlab-runner/aws"
  version = "5.9.1"

  aws_region  = var.aws_region
  environment = var.environment

  vpc_id                   = module.vpc.vpc_id
  subnet_ids_gitlab_runner = module.vpc.private_subnets
  subnet_id_runners        = element(module.vpc.private_subnets, 0)

  runners_name                     = var.runner_name
  runners_gitlab_url               = var.gitlab_url
  enable_runner_ssm_access         = true
  enable_docker_machine_ssm_access = true
  docker_machine_instance_type     = var.docker_machine_instance_type
  instance_type                    = var.instance_type
  cache_bucket_versioning          = true
  ami_filter                       = var.ami_filter
  runner_ami_filter                = var.runner_ami_filter
  docker_machine_download_url = var.docker_machine_download_url
    docker_machine_options = [
      "engine-install-url='https://releases.rancher.com/install-docker/20.10.21.sh'"
    ]

  runners_idle_count          = 1

  gitlab_runner_registration_config = {
    registration_token = var.registration_token
    tag_list           = "docker_spot_runner"
    description        = "runner default"
    locked_to_project  = "true"
    run_untagged       = "false"
    maximum_timeout    = "3600"
  }
}
```

**AWS Clouwatch error:**

```json
Feb 15 16:09:17 ip-10-0-1-76 gitlab-runner: {
    "driver": "amazonec2",
    "level": "info",
    "msg": "(runner-4zt1ivvv-runners-dev-arm6-1676477356-67e79255) Missing instance ID, this is likely due to a failure during machine creation",
    "name": "runner-4zt1ivvv-runners-dev-arm6-1676477356-67e79255",
    "operation": "create",
    "time": "2023-02-15T16:09:17Z"
}

Feb 15 16:09:19 ip-10-0-1-76 gitlab-runner: {
    "level": "error",
    "msg": "InvalidInstanceID.Malformed: The instance ID '' is malformed",
    "name": "runner-4zt1ivvv-runners-dev-arm6-1676477356-7981f938",
    "operation": "stop",
    "time": "2023-02-15T16:09:19Z"
}
```

**Manually create docker-machine instances from GitLab Runner Manager:**

```bash
sh-4.2$ docker-machine create --driver amazonec2 --amazonec2-region eu-central-2 aws01
Creating CA: /home/ssm-user/.docker/machine/certs/ca.pem
Creating client certificate: /home/ssm-user/.docker/machine/certs/cert.pem
Running pre-create checks...
Creating machine...
(aws01) Launching instance...
(aws01) Missing instance ID, this is likely due to a failure during machine creation
Error creating machine: Error in driver during machine creation: UnauthorizedOperation: You are not authorized to perform this operation. Encoded authorization failure message: vA9jaFqSW2xOiXdCN4ukdel2HbKtFTTMjnbYH4HaObZGlwEicaGhKDTGIAyjCS0ImKiiLYqcNhyf9ttbMQIyF8t4KPLl_jqFa8SnBAadfyKtBTlKo-hV-xZwPuvJ0NZ6ngeIlADk0f8YMUGXB1b7YQXCjyRt5wXb2q-41qrG8pK2txxcP6U_xc0Mb8EiSdyJyE_4f0Ft-b71zQ0kLKp4RL8F4Fr-GklhPbazRFUE0dEp2xR764IudqQpV5Sx-YAlCIxSoUZ6SlPSED4q3JHcS18jaU1UkMibNrAhbwKZN2A9WfkRusyGIGUotxgJZfI4Oqq0xyZXWGKpXGrd4aPz2KwEp7OxsCskbzdhXYW4S046EIkPOVhamqLO9OXrNgAEMXKFEP9JiP1EBIm2E0GAEtvEGYi7M0vAWnGT0shcSVjs2_2tlrD4hq1Xjc4rsEQLuvBxc6WbryO662erNyDDbZQuDHnFlTFmJqApqKw8TyrNfyBJ5zXEzZ4xiioOkVcEnW0lsAKb9woFcPBSC5vM2MDhVy02CmA93gELXRgshWUohOeB5NGz-7WR5iX5B8kI0mD4Qm9boDK3xOrQ2LQdmopQHXna3Vh9X3O5X_P73Q
        status code: 403, request id: 2413edc9-1809-494f-81df-d1f6f0e33b65
```

**Comment by kayman-mk** [comment]
Spot instances have a problem with the newest docker+machine driver. See https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/100

It works with
```hcl
version = "5.9.1"

gitlab_runner_version = "15.1.1"
docker_machine_options           = ["engine-install-url=https://releases.rancher.com/install-docker/20.10.21.sh"]
  docker_machine_version           = "0.16.2-gitlab.15"
```

**Comment by kayman-mk** [comment]
I close this one in favor of #696. There are working configurations available.
